common/cpp/src/google_smart_card_common/value.h

```diff
@@ -0,0 +1,137 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_SMART_CARD_COMMON_VALUE_H_
+#define GOOGLE_SMART_CARD_COMMON_VALUE_H_
+
+#include <stdint.h>
+
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace google_smart_card {
+
+// A variadic data type that approximately corresponds to JSONifiable types.
+//
+// Is intended to be used in generic interfaces related to message exchanging
+// with remote callers/receivers, for instance, for sending/receiving messages
+// to/from JavaScript code.
+//
+// Inspired by Chromium's base::Value, Pepper's pp::Var classes and JavaScript
+// type system.
+class Value final {
+ public:
+  enum class Type {
+    kNull,
+    kBoolean,
+    kInteger,
+    kFloat,
+    kString,
+    kBinary,
+    kDictionary,
+    kArray,
+  };
+
+  using BinaryStorage = std::vector<uint8_t>;
+  using DictionaryStorage = std::map<std::string, std::unique_ptr<Value>>;
+  using ArrayStorage = std::vector<std::unique_ptr<Value>>;
+
+  Value();
+  explicit Value(Type type);
+  explicit Value(bool boolean_value);
+  explicit Value(int integer_value);
+  explicit Value(int64_t integer_value);
+  explicit Value(double float_value);
+  explicit Value(const char* string_value);
+  explicit Value(std::string string_value);
+  explicit Value(BinaryStorage binary_value);
+  explicit Value(DictionaryStorage dictionary_value);
+  explicit Value(ArrayStorage array_value);
+  Value(Value&& other);
+  Value& operator=(Value&& other);
+  Value(const Value&) = delete;
+  Value& operator=(const Value&) = delete;
+  ~Value();
+
+  Type type() const { return type_; }
+  bool is_null() const { return type_ == Type::kNull; }
+  bool is_boolean() const { return type_ == Type::kBoolean; }
+  bool is_integer() const { return type_ == Type::kInteger; }
+  bool is_float() const { return type_ == Type::kFloat; }
+  bool is_string() const { return type_ == Type::kString; }
+  bool is_binary() const { return type_ == Type::kBinary; }
+  bool is_dictionary() const { return type_ == Type::kDictionary; }
+  bool is_array() const { return type_ == Type::kArray; }
+
+  bool GetBoolean() const;
+  int64_t GetInteger() const;
+  double GetFloat() const;
+  const std::string& GetString() const;
+  const BinaryStorage& GetBinary() const;
+  const DictionaryStorage& GetDictionaryItems() const;
+  const ArrayStorage& GetArrayItems() const;
+
+  const Value* GetDictionaryItem(const std::string& key) const;
```

Comment:
nit: I think this could use a comment about it returning nullptr when the key is not present